api/routes/model.js

```diff
@@ -501,6 +502,144 @@ router.put(
     res.json({ success: `${stageName} description updated!` });
   }
 );
+router.get("/:model_ID/canEdit", checkToken, async function(req, res) {
+  const db = req.db;
+  const collection = db.get("projects");
+  const { model_ID } = req.params;
+
+  const userEmail = req.decoded.sub;
+  const userId = await getUserId(db, userEmail);
+
+  collection
+    .findOne({
+      _id: model_ID,
+      ownerId: userId
+    })
+    .then(model =>
+      model !== null ? res.json({ success: true }) : res.sendStatus(404)
+    )
+    .catch(() => res.sendStatus(500));
+});
+
+router.post("/:model_ID", checkToken, async (req, res) => {
+  const db = req.db;
+  const collection = db.get("projects");
+  const { model_ID } = req.params;
+  const { name, description, groupSize } = req.body;
+
+  const userEmail = req.decoded.sub;
+  const userId = await getUserId(db, userEmail);
+
+  collection
+    .findOneAndUpdate(
+      {
+        _id: model_ID,
+        ownerId: userId
+      },
+      { $set: { name, description, groupSize } }
+    )
+    .then(model =>
+      model !== null
+        ? res.json({ success: `${name} updated!` })
+        : res.sendStatus(404)
+    )
+    .catch(() => res.sendStatus(500));
+});
+
+router.post("/:model_ID/:phaseName/:stageName", checkToken, async function(
+  req,
+  res
+) {
+  const db = req.db;
+  const collection = db.get("projects");
+  const { model_ID, phaseName, stageName } = req.params;
+
+  const { startdate, enddate } = req.body;
+  if (startdate === undefined || enddate === undefined) {
+    res.sendStatus(400);
+    return;
+  }
+
+  const userEmail = req.decoded.sub;
+  const userId = await getUserId(db, userEmail);
+
+  const newStage = { name: stageName, description: "", startdate, enddate };
+
+  collection
+    .findOneAndUpdate(
+      {
+        _id: model_ID,
+        ownerId: userId
+      },
+      { $push: { [`phases.${phaseName}.stages`]: newStage } }
+    )
+    .then(model =>
+      model !== null
+        ? res.json({ success: `${stageName} added!` })
+        : res.sendStatus(404)
+    )
+    .catch(() => res.sendStatus(500));
+});
+
+router.put(
+  "/:model_ID/:phaseName/:stageName/description",
+  checkToken,
+  async function(req, res) {
+    const db = req.db;
+    const collection = db.get("projects");
+    const { model_ID, phaseName, stageName } = req.params;
+    const description = req.body.description;
+    const subDescription = req.body.subDescription;
+    if (description === undefined) {
+      res.sendStatus(400);
+    }
+    collection
+      .findOneAndUpdate(
+        {
+          _id: model_ID,
+          [`phases.${phaseName}.stages.name`]: stageName
+        },
+        { $set: { [`phases.${phaseName}.stages.$.description`]: description } }
+      )
+      .then(model => {
+        if (model === null) {
+          res.sendStatus(404);
+        }
+      })
+      .catch(() => res.sendStatus(500));
```

Comment:
this request might send a 404/500 and still continue with the second part, which is both not what we want and would cause a server error due to multiple response attempts

maybe try converting this to async/await syntax so that we can break out of error results and return immediately

Reply:
wait i wrote lines 184-206 of user.js HAHA